src/NLog/Targets/DatabaseTarget.cs

```diff
@@ -629,32 +664,46 @@ protected override void Write(IList<AsyncLogEventInfo> logEvents)
-        private void WriteEventToDatabase(LogEventInfo logEvent)
+        private void WriteEventToDatabase(IDbCommand command, LogEventInfo logEvent, Layout commandLayout, IList<DatabaseParameterInfo> databaseParameterInfos, bool prepareDbCommand)
         {
-            //Always suppress transaction so that the caller does not rollback logging if they are rolling back their transaction.
-            using (TransactionScope transactionScope = new TransactionScope(TransactionScopeOption.Suppress))
-            {
-                EnsureConnectionOpen(BuildConnectionString(logEvent));
-
-                using (IDbCommand command = _activeConnection.CreateCommand())
-                {
-                    command.CommandText = RenderLogEvent(CommandText, logEvent);
-                    if (command.CommandType != CommandType)
-                    {
-                        command.CommandType = CommandType;  // Some DbProviders will throw when trying to modify CommandType
-                    }
+            var commandText = RenderLogEvent(commandLayout, logEvent);
 
-                    InternalLogger.Trace("DatabaseTarget(Name={0}): Executing {1}: {2}", Name, command.CommandType, command.CommandText);
+            InternalLogger.Trace("DatabaseTarget(Name={0}): Executing {1}: {2}", Name, command.CommandType, commandText);
 
-                    AddParametersToCommand(command, Parameters, logEvent);
-
-                    int result = command.ExecuteNonQuery();
-                    InternalLogger.Trace("DatabaseTarget(Name={0}): Finished execution, result = {1}", Name, result);
+            if (databaseParameterInfos.Count > 0)
+            {
+                bool createParameters = !prepareDbCommand || command.Parameters.Count == 0;
+                for (int i = 0; i < databaseParameterInfos.Count; ++i)
+                {
+                    var parameterInfo = databaseParameterInfos[i];
+                    var dbParameter = createParameters ? CreateDatabaseParameter(command, parameterInfo) : (IDbDataParameter)command.Parameters[i];
+                    var dbParameterValue = CreateDatabaseParameterValue(logEvent, parameterInfo);
+                    dbParameter.Value = dbParameterValue;
+                    if (createParameters)
+                        command.Parameters.Add(dbParameter);
+                    InternalLogger.Trace("  DatabaseTarget: Parameter: '{0}' = '{1}' ({2})", dbParameter.ParameterName, dbParameter.Value, dbParameter.DbType);
                 }
+            }
 
-                //not really needed as there is no transaction at all.
-                transactionScope.Complete();
+            if (!prepareDbCommand || command.CommandText != commandText)
+            {
+                command.CommandText = commandText;
+                if (prepareDbCommand)
+                {
+                    command.Prepare();
```

Comment:
I'm not sure if we should prepare every sql query (worst case).

I would expect this in the init method.

Reply:
Could make a cache of prepared DbCommands, and perform lookup using the key of ConnectionString + CommandText.